File: study/src/Patterns/proxy/SpaceShip.java
package Patterns.proxy;

import Patterns.Display;
import Patterns.Drawable;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class SpaceShip extends Drawable implements Runnable {

    private Color color;
    private int radius;
    private boolean isFlying = false;
    private double speed = 15;

    private Planet destination;
    private double x;
    private double y;
    private static final double capacity = 20;
    private double load = 0;
    private String loadDescription;
    List<Planet> destinations =  new ArrayList<Planet>();


    public SpaceShip(double x, double y, int radius, int r, int g, int b) {
        this.x = x;
        this.y = y;
        this.radius = radius;
        this.color = new Color(r, g, b);
    }

    @Override
    public void draw() {
        Graphics graphics = Display.getInstance().getGraphics();
        // draw the door
        graphics.setColor(color);
        graphics.fillArc((int)(x - radius/2), (int)(y - radius), radius, radius*2, 0, 360);
    }

    public void addPointToRoute(Planet planet) {
        destinations.add(planet);
    }

    public void flyToNextPoint() throws ShipHasNoDestinationException {
        if(destinations.isEmpty()) {
            throw new ShipHasNoDestinationException();
        }

        setCourse(destinations.remove(0));
    }

    public void undock() throws ShipHasNoDestinationException {
        flyToNextPoint();
    }


    public void setCourse(Planet destination) {
        this.destination = destination;
        isFlying = true;
    }

    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
//            System.out.println("Ship load: " + load);
            if (isFlying) {
                if (Math.abs(x - destination.getPx()) <= speed && Math.abs(y - destination.getPy()) <= speed )  {
                    x = destination.getPx();
                    y = destination.getPy();
                    isFlying = !destination.dock(this);
                } else {
                    double dX = destination.getPx() - x;
                    double dY = destination.getPy() - y;

                    double k = Math.sqrt(dX * dX + dY * dY)/speed;
                    double vX = dX/k;
                    double vY = dY/k;

                    x = x + vX;
                    y = y + vY;

                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                        break;
                    }
                }
            }
        }
    }

    public void batchLoad(double batch) throws ShipOverloadedException {
        if(load + batch <= capacity) {
            load += batch;
        } else {
            throw new ShipOverloadedException();
        }
    }

    public double batchUnload(double batch) {
        if (load < batch) {
            double temp = load;
            load = 0;
            return temp;
        } else {
            load = load - batch;
            return batch;
        }
    }

    public double getFreeCapacity() {
        return capacity - load;
    }

    public double getLoad() {
        return load;
    }
}

File: study/src/kt/proxy_k/SpaceShip.kt
package kt.proxy_k

import Patterns.Display
import Patterns.proxy.ShipHasNoDestinationException
import Patterns.proxy.ShipOverloadedException
import java.awt.Color
import kotlin.math.abs
import kotlin.math.sqrt

class SpaceShip(
    x: Double,
    y: Double,
    var radius: Int,
    var r: Int,
    var g: Int,
    var b: Int,
    private var isFlying: Boolean = false,
    private var color: Color = Color(r, g, b),
    private var speed: Int = 15,
    private var destination: Planet,
    private val capacity: Int = 20,
    private var load: Int = 0,
    private var destinations: ArrayList<Planet> = ArrayList()
): Drawable(x, y), Runnable {

    override fun draw() {
        val graphics = Display.getInstance().graphics
        graphics.color = color
        graphics.fillArc((x - radius/2).toInt(), (y - radius).toInt(), radius, radius * 2, 0, 360)
    }

    fun addPointToRoute(planet: Planet) {
        destinations.add(planet)
    }

    fun flyToNextPoint() {
        if (destinations.isEmpty()) {
            throw ShipHasNoDestinationException()
        }
        setCourse(destinations.removeAt(0))
    }

    fun undock() {
        flyToNextPoint()
    }

    fun setCourse(incomeDestination: Planet) {
        destination = incomeDestination
        isFlying = true
    }

    fun getFreeCapacity() = capacity - load

    fun batchLoad(batch: Int) {
        if (load + batch <= capacity) {
            load +=batch
        } else throw ShipOverloadedException()
    }

    fun batchUnload(batch: Int): Int {
        if (load < batch) {
            val temp = load
            load = 0
            return temp
        } else {
            load -= batch
            return batch
        }
    }

    fun getLoad() = load


    override fun run() {
        while (!Thread.currentThread().isInterrupted) {
                if (isFlying) {
                    if (abs(x - destination.x) <= speed && abs(y - destination.y) <= speed) {
                        x = destination.x
                        y = destination.y
                        isFlying = !destination.dock(this)
                    }
                    else {
                        val dX = destination.x - x
                        val dY = destination.y - y

                        val k = sqrt((dX*dX + dY*dY).toDouble())/speed
                        val vX = dX/k
                        val vY = dY/k

                        x += vX
                        y += vY

                        try {
                            Thread.sleep(100)
                        } catch (e: InterruptedException) {
                            break
                        }

                    }
                }
            }
    }

}
